File: javadarts/src/test/java/com/sdoward/barstewarddarts/android/java/TurnValidatorTest.java
package com.sdoward.barstewarddarts.android.java;

import org.junit.Test;

import static org.junit.Assert.*;
import static org.assertj.core.api.Assertions.*;


public class TurnValidatorTest {

    private final TurnValidator turnValidator = new TurnValidator();

    @Test
    public void shouldReturnFalseWhenTurnIsNotComplete() {

        assertThat(turnValidator.turnIsComplete()).isFalse();

        turnValidator.acceptThrow(Throw.innerBull());

        assertThat(turnValidator.turnIsComplete()).isFalse();

        turnValidator.acceptThrow(Throw.innerBull());

        assertThat(turnValidator.turnIsComplete()).isFalse();
    }

    @Test
    public void shouldReturnTrueWhenTurnIsComplete() {
        turnValidator.acceptThrow(Throw.innerBull());
        turnValidator.acceptThrow(Throw.outerBull());
        turnValidator.acceptThrow(Throw.miss());

        assertThat(turnValidator.turnIsComplete()).isTrue();
    }

    @Test(expected = IllegalStateException.class)
    public void shouldThrowExceptionWhenForthTurnIsAdded() {
        turnValidator.acceptThrow(Throw.innerBull());
        turnValidator.acceptThrow(Throw.outerBull());
        turnValidator.acceptThrow(Throw.miss());
        turnValidator.acceptThrow(Throw.duble(1));
    }

    @Test
    public void shouldReturnTurnWhenValid() {

        turnValidator.acceptThrow(Throw.innerBull());
        turnValidator.acceptThrow(Throw.outerBull());
        turnValidator.acceptThrow(Throw.miss());

        Turn expectedTurn = Turn.create(Throw.innerBull(), Throw.outerBull(), Throw.miss());
        Turn actualTurn = turnValidator.getTurn();

        assertThat(expectedTurn).isEqualTo(actualTurn);
    }


}

File: kotlindarts/src/test/kotlin/com/sdoward/barstewarddarts/android/kotlin/TurnValidatorTest.kt
package com.sdoward.barstewarddarts.android.kotlin

import org.assertj.core.api.Assertions.assertThat
import org.junit.Test

class TurnValidatorTest {

    private val turnValidator = TurnValidator

    @Test
    fun shouldReturnFalseWhenTurnIsNotComplete() {
        assertThat(turnValidator.turnIsComplete()).isFalse()

        turnValidator.acceptThrow(Throw.innerBull())

        assertThat(turnValidator.turnIsComplete()).isFalse()

        turnValidator.acceptThrow(Throw.innerBull())

        assertThat(turnValidator.turnIsComplete()).isFalse()
    }

    @Test
    fun shouldReturnTrueWhenTurnIsComplete() {
        turnValidator.resetTurn() // not sure why this is necessary

        turnValidator.acceptThrow(Throw.innerBull())
        turnValidator.acceptThrow(Throw.outerBull())
        turnValidator.acceptThrow(Throw.miss())

        assertThat(turnValidator.turnIsComplete()).isTrue()
    }

    @Test(expected = IllegalStateException::class)
    fun shouldThrowExceptionWhenForthTurnIsAdded() {
        turnValidator.resetTurn()

        turnValidator.acceptThrow(Throw.innerBull())
        turnValidator.acceptThrow(Throw.outerBull())
        turnValidator.acceptThrow(Throw.miss())
        turnValidator.acceptThrow(Throw.duble(1))
    }

    @Test
    fun shouldReturnTurnWhenValid() {

        turnValidator.acceptThrow(Throw.innerBull())
        turnValidator.acceptThrow(Throw.outerBull())
        turnValidator.acceptThrow(Throw.miss())

        val expectedTurn = Turn(Throw.innerBull(), Throw.outerBull(), Throw.miss())
        val actualTurn = turnValidator.turn

        assertThat(expectedTurn).isEqualTo(actualTurn)
    }

}